crates/runtime/src/continuation.rs

```diff
@@ -418,14 +432,20 @@ pub fn resume(
 /// TODO
 #[inline(always)]
 pub fn suspend(instance: &mut Instance, tag_index: u32) {
-    let stack_ptr = TopOfStackPointer::as_raw(instance.tsp());
-    let parent = unsafe { stack_ptr.cast::<*mut u8>().offset(-2).read() };
+    let running = instance.typed_continuations_store();
+    let running = unsafe {
+        running
+            .as_ref()
+            .expect("Calling suspend outside of a continuation")
```

Comment:
I guess we can store the suspend arguments directly on the parent continuationobject instead of mangling with the `VMContext` now.

Reply:
We should generally improve error handling, in particular trapping with an appropriate error code.

Regarding storing suspend arguments in continuation objects: You are right, the fact that we now have easy access to the currently running continuation object allows us to store the suspend arguments in there. However, there are a series of efficiency trade-offs: The "global" buffer currently used for suspend arguments can be re-used between different suspends, only being re-allocated whenever it is too small to hold the necessary number of suspend arguments. It also makes the current (rather naive) implementation of effect forwarding more efficient: While buffers local to continuation objects have to be handed over from parent to child (see the `cont_obj_forward_tag_return_values_buffer` builtin), this is not the case with the global payload buffer in the `VMContext`.

In total, I would suggest that we continue using the "global" buffer in the `VMContext` for suspend arguments for the time being, and revisit a better design once we either optimised effect forwarding or implement some kind of pooling logic for continuation objects and/or the buffers therein.

Reply:
I think it would be good to not mangle with the `VMContext` too much. I agree that it would be more inefficient in the case of forwarding, but that's only because forwarding is implemented in a particularly naive way right now. Thus I don't worry about that at all.

Reply:
Just to be clear, I am not suggesting we do this refactor as part of this PR.